Add path/4 for storage backends

fastjames opened the pull request:
In order to make the object path available from the S3 backend (e.g. for
direct access via ex-aws-s3), refactor path logic into a new public
function.

fastjames commented:
> Hey,
> 
> > (e.g. for direct access via ex-aws-s3),
> 
> Could you please describe what you mean by this? Currently you can take url result to extract path.
> 
> Regarding the local storage, waffle has file path relative to root directory and public's file path (excluding `storage_dir_prefix` prefix), it's not clear whether path should include `storage_dir_prefix` from description of your issue.
> 
> Thank you

Hello, thanks for your feedback. I made the change to local storage thinking that I needed to maintain parity, but I can see how it might be ambiguous in this case.

For S3 storage, what I wanted was the object key within the bucket. I have a function that creates a line-oriented stream for an S3 object, so I feed the object key to that function.

Given that S3 URLs can take multiple forms where the bucket name is part of the path, or maybe part of the host, it seemed risky to write my own logic for chopping up the URL. I could see that the actual object key was already being used here, so refactoring it into its own function made the most sense to me at the time.

achempion commented:
ok, I see now 👍 

I'll mere it soon, it would be nice to have a related test cases and maybe a documentation for a new public function

fastjames commented:
> ok, I see now 👍
> 
> I'll merge it soon, it would be nice to have a related test cases and maybe a documentation for a new public function

Ah, a test would be a great addition. I will write those and update the PR branch.

fastjames commented:
I added a simple test and some macro code to make it pass. I also included some instructions on how to run the test suite, and I added an env var way to set the S3 region. I'm happy to re-section the PR into atomic commits if that's more in line with the repo history.

achempion commented on lib/waffle/storage/local.ex:
```diff
@@ -42,20 +42,24 @@ defmodule Waffle.Storage.Local do
     |> URI.encode()
   end
 
+  def path(definition, version, file_and_scope) do
```
if it returns path in storage, we probably should include prefix path for local storage

fastjames replied:
That's a good point. I have updated `Local.path/3` to behave more like `url/3` while never adding the asset host. That should produce a usable path.

achempion commented on lib/waffle/storage/s3.ex:
```diff
@@ -159,6 +159,11 @@ defmodule Waffle.Storage.S3 do
 
+  def path(definition, version, file_and_scope, _options \\ []) do
+    s3_key(definition, version, file_and_scope)
+    |> Url.sanitize(:s3)
```
why we do sanitization if it's a path and not url?

achempion reviewed (commented):
We still don't respect `storage_dir_prefix` setting for local storage.

The overall result is a bit overcomplicated :) It repeats url in many ways, but url generation is overcomplicated as well :)

Maybe we can just make `s3_key` public and I believe you can reference public adapter function through the `_storage` attribute.

fastjames commented:
> We still don't respect `storage_dir_prefix` setting for local storage.
> 
> The overall result is a bit overcomplicated :) It repeats url in many ways, but url generation is overcomplicated as well :)
> 
> Maybe we can just make `s3_key` public and I believe you can reference public adapter function through the `_storage` attribute.

I like simplicity too, so I opened a PR based on your recommendation at https://github.com/elixir-waffle/waffle/pull/100 . I believe that simpler change will meet my needs without adding more complexity.
